Refactor Python CUDA IPC API

Mivik opened the pull request:
As noted in #978, the current Python CUDA IPC API fails when sending specific tensors.

My investigation suggests that `pyarrow.cuda` (used by `dora.cuda`), is likely the source of this failure. During testing with two processes (sender and receiver), the receiver receives `pyarrow.CudaBuffer` with corrupted content, and it appears that there's nothing we can do about this.

However, I found that the `numba.cuda` module is a suitable alternative for implementing CUDA IPC. As we already depend on `numba`, no new dependencies are introduced. I have refactored the API to use `numba.cuda`, and it has passed the tests proposed in #978.

Example & docstrings are updated accordingly.

**Note**: this change is not backward compatible.

haixuanTao commented:
Thanks for this investigation and this looks really cool! 

However, would it be possible to not depend on pickle as it would reduce GPU 0-copy to python only project.

The idea behind using cuda ipc handle was to be able to read this in like C/C++ or Rust.

Mivik commented:
> However, would it be possible to not depend on pickle as it would reduce GPU 0-copy to python only project.

I refactored the encoding part and now removed pickle.

haixuanTao commented on apis/python/node/dora/cuda.py:
```diff
@@ -24,75 +30,65 @@ def torch_to_ipc_buffer(tensor: torch.TensorType) -> tuple[pa.array, dict]:
+        "source_info": json.dumps(source_info),
     }
-    return pa.array(handle_buffer, type=pa.uint8()), metadata
+    return pa.array(handle, pa.int8()), metadata
```
This should be uint8 right?
```suggestion
    return pa.array(handle, pa.uint8()), metadata
```

Mivik replied:
`handle` is actually a int8 array (ranged from `-128` to `127`), since its underlying c type is `char`

haixuanTao reviewed (approved):
Looks good to me!

I think uint8 would be more resilient!

But thanks for everything!
